src/backend/distributed/planner/insert_select_planner.c

```diff
@@ -802,7 +744,18 @@ DistributedInsertSelectSupported(Query *queryTree, RangeTblEntry *insertRte,
 	}
 	else if (IsCitusTableType(targetRelationId, NULL_KEY_DISTRIBUTED_TABLE))
 	{
-		/* we've already checked the subquery via ErrorIfInsertSelectWithNullDistKeyNotSupported */
+		/*
+		 * Note that we've already checked the non-existence of Postgres
+		 * tables in the subquery.
+		 */
+		if (subqueryRteListProperties->hasCitusLocalTable ||
+			subqueryRteListProperties->hasMaterializedView)
+		{
+			return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+								 "cannot select from a local relation when inserting "
+								 "into a distributed table that does not have a shard "
+								 "key", NULL, NULL);
```

Comment:
It would be nicer if we have this check for distributed tables in general, see https://github.com/citusdata/citus/pull/6817.